File: src-tauri/src/models/response.model.rs
/* sys lib */
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub enum ResponseStatus {
  Success,
  Info,
  Warning,
  Error,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum DataValue {
  String(String),
  Number(f64),
  Bool(bool),
  Array(Vec<serde_json::Value>),
  Object(serde_json::Value),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ResponseModel {
  pub status: ResponseStatus,
  pub message: String,
  pub data: DataValue,
}

impl From<Box<dyn std::error::Error + Send + Sync>> for ResponseModel {
  fn from(error: Box<dyn std::error::Error + Send + Sync>) -> Self {
    ResponseModel {
      status: ResponseStatus::Error,
      message: error.to_string(),
      data: DataValue::String("".to_string()),
    }
  }
}

impl From<serde_json::Error> for ResponseModel {
  fn from(error: serde_json::Error) -> Self {
    ResponseModel {
      status: ResponseStatus::Error,
      message: error.to_string(),
      data: DataValue::String("".to_string()),
    }
  }
}

impl From<String> for ResponseModel {
  fn from(error: String) -> Self {
    ResponseModel {
      status: ResponseStatus::Error,
      message: error,
      data: DataValue::String("".to_string()),
    }
  }
}

File: src/app/models/response.model.ts
export enum ResponseStatus {
  SUCCESS = "Success",
  INFO = "info",
  WARNING = "Warning",
  ERROR = "Error",
}

export enum ResponseStatusIcon {
  SUCCESS = "check_circle_outlined",
  INFO = "info_outlined",
  WARNING = "warning_outline",
  ERROR = "dangerous",
  "" = "notifications",
}

export interface INotify {
  status: ResponseStatus;
  message: string;
}

export interface ActiveNotification extends INotify {
  id: number;
  icon: ResponseStatusIcon;
  intervalId?: number;
  timeoutId?: number;
}

export interface Response<T> {
  status: ResponseStatus;
  message: string;
  data: T;
}
